Orian bioconductor320 updates

orian116 opened the pull request:
Updated SpaceMarkersMetric by fixing signage and log transformed to scale
magnitude
Added getSpatialParameters_External which enables getting spatial parameters
from file or from the user
Enabled includeSelf = TRUE in getInteractingGenes.R to improve hotspot
detection
Enabled load10XCoords to read coordinates from VisiumHD directory
NB: load10XExpr does not need to be explicitly updated since you can just supply visiumDir as in regular Visium

dimalvovs commented on R/preprocessing.R:
```diff
@@ -116,15 +125,21 @@ load10XCoords <- function(visiumDir, resolution = "lowres", version = NULL){
         has_header <- TRUE
         tissue_pos_name <- "tissue_positions.csv"
     }
```
would it make sense to continue adding version treatments here? for example `..else if (version == "HD")...`

dimalvovs replied:
and add respective coord readers for example, to avoid ifs downstream, like `coord_reader <- read.csv` for versions 1 and 2, and `coor_reader <- nanoparquet::read_parquet` for HD

orian116 replied:
Hi Dima, Do you mean like this?
load10XCoords <- function(visiumDir, resolution = "lowres", version = NULL){
    #determine spacerager version
    if(is.null(version)){
        message("Version not provided. Trying to infer.")
        if("probe_set.csv" %in% dir(visiumDir)){
            config_line <- readLines(paste0(visiumDir,"/probe_set.csv"), 1)
            version <- strsplit(config_line, "=")[[1]][2]
        } else if ("tissue_positions.parquet" %in% dir(
          visiumDir,"binned_outputs/square_008um/spatial")) {
          version <- "HD"
          visiumDir <- file.path(visiumDir,"binned_outputs/square_008um")
          message(".parquet file found. 
                  Assuming VisiumHD with 008um resolution as default")
        } else {
            message("probe_set.csv  or .parquet not found. 
                    Assuming version 1.0.")
            version <- "1.0"
        }
    }
    spatial_dir <- paste0(visiumDir,"/spatial")
    #account for different versions of visium data
    if(version == "1.0"){
        has_header <- FALSE
        tissue_pos_name <- "tissue_positions_list.csv"
        coord_file <- dir(spatial_dir,
                          pattern = tissue_pos_name, full.names = TRUE)
        coord_values <- read.csv(coord_file, header = has_header)
    } else if (version == "2.0") {
        has_header <- TRUE
        tissue_pos_name <- "tissue_positions.csv"
        coord_file <- dir(spatial_dir,
                          pattern = tissue_pos_name, full.names = TRUE)
        coord_values <- read.csv(coord_file, header = has_header)
    } else if (version == "HD") {
      tissue_pos_name <- "tissue_positions.parquet"
      coord_file <- dir(spatial_dir,pattern = tissue_pos_name,
                        full.names = TRUE)
      coord_values <- nanoparquet::read_parquet(coord_file)
    }
    scale_json <- dir(spatial_dir,
                        pattern = "scalefactors_json.json",full.names = TRUE)
    scale_values <- jsonlite::read_json(scale_json)
    scale_factor <- scale_values[grepl(resolution, names(scale_values))][[1]]
    coord_values <- coord_values[,c(1,5,6)]
    coord_values[,2:3] <- coord_values[,2:3]*scale_factor
    names(coord_values) <- c("barcode","y","x")
    return(coord_values)
}

dimalvovs replied:
oh that could work too as all switches would be in one place. 

Initially I thought of a much shorter
```
if(version == "1.0"){
  has_header <- FALSE
  tissue_pos_name <- "tissue_positions_list.csv"
  coord_reader <- read.csv
} else if (version == "2.0") {
  has_header <- TRUE
  tissue_pos_name <- "tissue_positions.csv"
  coord_reader <- read.csv
} else if (version == "HD") {
  has_header <- TRUE
  tissue_pos_name <- "tissue_positions.parquet"
  coord_reader <- nanoparquet::read_parquet
}
```

followed by 

```
coord_file <- dir(spatial_dir,pattern = tissue_pos_name, full.names = TRUE)
if(!is.null(formals(coord_reader)["header"])) formals(coord_reader)["header"]
coord_values <- coord_reader(coord_file)
```
but maybe the use of formals is a bit opaque.

dimalvovs replied:
thank you for amending the code, and sorry for not spotting this at first sight - would we benefit form a test that checks visium HD data loading? Like is done [here](https://github.com/DeshpandeLab/SpaceMarkers/blob/orian_Bioconductor3-20_updates/tests/testthat/test-preprocessing.load10XCoords.R)

dimalvovs commented on R/preprocessing.R:
```diff
@@ -103,8 +105,15 @@ load10XCoords <- function(visiumDir, resolution = "lowres", version = NULL){
             version <- strsplit(config_line, "=")[[1]][2]
+        } else if ("tissue_positions.parquet" %in% dir(
+          visiumDir,"binned_outputs/square_008um/spatial")) {
+          version <- "HD"
```
if we've added support for HD I think we should proudly shout about it in the news

dimalvovs replied:
my bad, it is there!

orian116 commented:
Thanks Dima. I updated the code to reflect your suggested comments for load10xCoords and I added tests for getSpatialParametersExternal. Should I close this pull request and create a new one?

dimalvovs commented:
Thank you! No need for the new pull request, your changes are reflected above. I'll have a look.
